### PR TITLE
Validate email address

### PR DESCRIPTION
**Pull request by McEileen**
Hi, please don't look at this code because it is not nearly done.  I just pushed to remote because I am having work done on my local machine tomorrow.  Thanks!

### CODE REVIEW

**Comment by McEileen** [comment]
Hi @javierforero, @vkoves or anyone else with devise experience (@alexstevenson19?) could one of you please look at this and tell me what I'm doing wrong with adding devise to the User model?  
When I try to create a new user through the sign-up form on my localhost, I get the below error
screen shot 2016-11-05 at 2 29 06 pm
I found this relevant github issue https://github.com/plataformatec/devise/issues/589.
 When I run `User.new.respond_to?(:encrypted_password=)` in `rails console`, I get `false`
The suggested solution of resetting the database, to ensure that the encrypted_password column is on User, didn't work 

I also checkout this question, which had a similar suggested solution. http://stackoverflow.com/questions/30269771/nomethoderror-in-deviseregistrationscontrollercreate-undefined-method-encry
Thank you !!!

**Comment by vkoves** [comment]
@McEileen I'm starting to take a look at this right now. In pulling to your branch and trying to run migrations I get some errors (in particular `PG::UniqueViolation: ERROR:  could not create unique index "index_users_on_encrypted_password"`) so I'm going to try and fix some stuff and see what I can do. I'll keep you posted and push to the branch if I make headway.

**Comment by vkoves** [comment]
@McEileen from some experimenting I found that commenting out the line in the migrations that adds an index by encrypted password (now `# add_index :users, :encrypted_password, unique: true`) fixes my inability to run the Devise migration. I'm not sure why this is, but worth documenting. With this in place, I actually get an error on signup (a Devise error, not a page breaking Rails error) that says "Name can't be blank".

There's no name field on the page, so this seems a little silly, so I'll be working on figuring this out as well, but I'd be curious to see if remigrating your database without the encryption index might fix your issue.

**Comment by vkoves** [comment]
Another update. I realized that `user.rb` actually has `validates :name, presence: true` (from the old system I presume), so I commented that out and sign up seems to work properly. I get redirected to the homepage and get shown this message:

> A message with a confirmation link has been sent to your email address. Please follow the link to activate your account.

However, due to mailing probably not being configured, I haven't gotten any email, but if you can replicate these fixes it should help you continue on with this feature. I haven't commited it since I'm not sure if it is unique to me, but if you'd like me to I can. Otherwise just try out removing that line in the migration that I mentioned earlier and removing name validation from the user model, and hopefully that should get you to at least sign up not failing.

**Comment by javierforero** [comment]
@vkoves good catch on the devise migration file. So it turns out the migration file its specifying ```encrypted_password``` to be unique, and it is not since the default value is an empty string ```""```. So rails freaks out cause it sees that there are duplicate values.  You can also fix this by changing the default value of ```encrypted_password``` to be ```nil```

@McEileen
In terms of the user name, don't we want that field for sign up? There is a way to add that to Devise's form. Here's a short tutorial [on it](http://jacopretorius.net/2014/03/adding-custom-fields-to-your-devise-user-model-in-rails-4.html)

**Comment by McEileen** [comment]
Thank you for all the help, @javierforero and @vkoves.  I got to the point @vkoves was at of being able to sign up successfully, but I also didn't get the confirmation email since we don't have that set up yet.  Besides the changes you all suggested, I also had to remove password_digest from the table, as I think it was leftover from the pre-devise authentication.

**Comment by McEileen** [comment]
I got email validation working use sendgrid, woot!  There are still some little things to fix before pushing, namely, the confirmation comes from an email called "please-change-me-at-config-initializers-devise@example.com" (LOL) and I need to safely store the sendgrid username/password in heroku.  Once that's done, I think this PR will be done.

**Comment by javierforero** [comment]
@McEileen awesome!!!

For keeping the credentials secure you could use the figaro gem. [Here's the getting started](https://github.com/laserlemon/figaro/blob/master/README.md )if ya need unless you have another gem that you like better.

**Comment by McEileen** [comment]
Hi @vkoves, can you please review my pull request?  Feedback on git hygiene, etc. etc., is welcome and wanted!

File: db/migrate/20160803021240_add_devise_to_users.rb [hunk 1]
```diff
@@ -38,7 +38,7 @@ def self.up
 
     add_index :users, :email,                unique: true
     add_index :users, :reset_password_token, unique: true
-    add_index :users, :encrypted_password, unique: true
+    # add_index :users, :encrypted_password, unique: true
```
Comment (vkoves) [inline db/migrate/20160803021240_add_devise_to_users.rb]: Typically migrations should not be changed after they are created, rather a new migration would be made to undo the changes made. However, this database is odd and mostly reads off the `schema.rb` anyway, so this isn't a big deal.

File: app/controllers/lessons_controller.rb [hunk 2]
```diff
@@ -47,42 +47,46 @@ def new
...
   def create
-    lesson = Lesson.new
-    lesson.hash_tag = params[:lesson][:hash_tag]
-    lesson.created_at = Time.now
-    lesson.category = params[:lesson][:category]
-    lesson.description = params[:lesson][:description]
-    lesson.approved = true
-    lesson.save
-    cookies["new_lesson_id"] = lesson.id
+    if user_signed_in?
```
Comment (vkoves) [inline app/controllers/lessons_controller.rb]: The rails `.new` action can take params and automatically put together params that are named like database columns. Refactor to use that and then do anything custom that needs it. See this [StackOverflow post](http://stackoverflow.com/questions/5667075/creating-rails-activerecord-model-from-params-hash). Looking at the diff I can see this was not made in this commit, but still worth pointing out.

File: app/controllers/lessons_controller.rb [hunk 3]
```diff
@@ -47,42 +47,46 @@ def new
...
+      lesson = Lesson.new
+      lesson.hash_tag = params[:lesson][:hash_tag]
+      lesson.created_at = Time.now
+      lesson.category = params[:lesson][:category]
+      lesson.description = params[:lesson][:description]
+      lesson.approved = true
+      lesson.save
+      cookies["new_lesson_id"] = lesson.id
```
Comment (vkoves) [inline app/controllers/lessons_controller.rb]: Why is a cookie set here? If this has a purpose it should be documented (as it seems a bit arbitrary), otherwise it should be removed. Also not in your work, but worth pointing out.

File: app/controllers/lessons_controller.rb [hunk 4]
```diff
@@ -47,42 +47,46 @@ def new
...
 
     # create a new contribution for creator
-    contribution = Contribution.new
-    contribution.lesson_id = lesson.id
-    contribution.user_id = session["user_id"]
-    contribution.creator = true
-    contribution.save
+      contribution = Contribution.new
+      contribution.lesson_id = lesson.id
+      contribution.user_id = session["user_id"]
```
Comment (vkoves) [inline app/controllers/lessons_controller.rb]: I believe this should be refactored to `current_user.id` which is more clear and uses the Devise helper for this.

File: app/controllers/lessons_controller.rb [hunk 5]
```diff
@@ -47,42 +47,46 @@ def new
...
+      contribution.creator = true
+      contribution.save
+    end
```
Comment (vkoves) [inline app/controllers/lessons_controller.rb]: This indentation is off, it should be corrected by ensuring this end lines up with the next line, and that everything in the if is indented inside that. It might just be the next line has too much indent.

File: app/controllers/lessons_controller.rb [hunk 6]
```diff
@@ -47,42 +47,46 @@ def new
...
-    @lesson.save
-
-    # create a new contribution for editor
-    contribution = Contribution.new
-    contribution.lesson_id = @lesson.id
-    contribution.user_id = session["user_id"]
-    contribution.creator = false
-    contribution.save
+    if user_signed_in?
+      @lesson.hash_tag = params[:lesson][:hash_tag]
```
Comment (vkoves) [inline app/controllers/lessons_controller.rb]: I again realize this was not part of your work, but this can also be refactored to use `.update_attributes`, which takes in params (see this [StackOverflow post](http://stackoverflow.com/questions/26090666/rails-update-object-from-params)). Also why is this `created_at` attribute modified? That should not be mutated after creation, otherwise it doesn't serve its purpose.

File: app/controllers/users/confirmations_controller.rb [hunk 7]
```diff
@@ -0,0 +1,28 @@
+class Users::ConfirmationsController < Devise::ConfirmationsController
+  # GET /resource/confirmation/new
```
Comment (vkoves) [inline app/controllers/users/confirmations_controller.rb]: Are these comments just to outline what paths exist? I'm not sure they are necessary. This seems to occur often, so I assume a command was run that generated the controllers, but any unused ones can be deleted, though they might be worth keeping.

File: app/controllers/users_controller.rb [hunk 8]
```diff
@@ -29,21 +29,21 @@ def new
...
-    user.updated_at = Time.now
-    user.save
-    #created = User.find_by(email: params[:user][:email])
-    #if !created
-    #  redirect_to users_url, notice: "Invalid E-mail"
-    #  return
-    #end
-    redirect_to lessons_url
-  end
+  # def create
```
Comment (vkoves) [inline app/controllers/users_controller.rb]: Remove this block of comments. Source control has this code if it needs to be reverted, and it adds bloat to have it in the source.

File: app/models/user.rb [hunk 9]
```diff
@@ -1,12 +1,14 @@
 class User < ActiveRecord::Base
+	
+	attr_accessor :username
```
Comment (vkoves) [inline app/models/user.rb]: Minor, but indentation seems off here.

File: app/views/devise/registrations/new.html.erb [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+
+
+  <div class="field">
+    <%= f.label :username %><br />
+    <%= f.text_field :username, autofocus: true %>
+  </div>
+
+  <div class="field">
+    <%= f.label :email %><br />
+    <%= f.email_field :email, autofocus: true %>
```
Comment (vkoves) [inline app/views/devise/registrations/new.html.erb]: It doesn't make much sense to have two fields using autofocus, remove one of the two, as I am unsure which one will actually receive focus on load.

File: app/views/devise/sessions/new.html.erb [hunk 11]
```diff
@@ -0,0 +1,26 @@
+<h2>Log in</h2>
+
+<%= form_for(resource, as: resource_name, url: session_path(resource_name)) do |f| %>
+  <div class="field">
+    <%= f.label :email %><br />
+    <%= f.email_field :email, autofocus: true %>
```
Comment (vkoves) [inline app/views/devise/sessions/new.html.erb]: It looks like the username is not requested on user creation. Wanted to point that out to check that it is intentional, as it is in the permitted fields for sign up.

File: app/views/shared/_navigation.html.erb [hunk 12]
```diff
@@ -28,13 +28,11 @@
...
-            <% current_user = User.find_by(id: session["user_id"]) %>
-            <li><%= link_to "Hello #{current_user.name} !", user_path(current_user), class:"btn btn-primary"%></li>
-            <li><%= link_to "Sign Out", logout_path, class:"btn btn-default"%></li>
+        <% if current_user %>
+            <%= link_to "Logout", destroy_user_session_path, method: :delete%>
         <% else %>
-            <li><%= link_to "Sign Up", new_user_path, class:"btn btn-default" %></li>
-            <li><%= link_to "Sign In", login_path, class:"btn btn-default" %></li>
+            <li><%= link_to "Sign Up", new_user_registration_path %></li>
+            <%= link_to "Login", new_user_session_path %>
```
Comment (vkoves) [inline app/views/shared/_navigation.html.erb]: It appears this line is not wrapped in an `<li>` tag and that both the Sign Up and Login buttons no longer have Bootstrap classes. Is that intentional?

File: app/views/users/new.html.erb [hunk 13]
```diff
@@ -1,3 +1,6 @@
+<!-- WE NEED TO USE THE DEVISE VIEWS FOR THE FORM -->
```
Comment (vkoves) [inline app/views/users/new.html.erb]: Devise creates the `registrations/new` view, which takes care of this for you, and should be the one your header link goes to. I'd suggest double checking that, and if it's all working, delete this view.

File: config/initializers/setup_mail.rb [hunk 14]
```diff
@@ -0,0 +1,10 @@
+ActionMailer::Base.delivery_method = :smtp
+ActionMailer::Base.smtp_settings = {
```
Comment (vkoves) [inline config/initializers/setup_mail.rb]: This data should probably not be in source control, I'd suggest using the SendGrid Heroku addon or using environment variables for this for security.

File: app/views/devise/sessions/new.html.erb (hunk 11 above)
Comment (McEileen) [inline app/views/devise/sessions/new.html.erb]: This is the user log-in, not registration, but you're right, this is definitely a bug.  I think that when I used devise to generate new views for registration and log-in, the views automatically included email but not username.  I fixed it in registration but not log-in.  I'll fix this now.  (Thanks for all the feedback!) 👍  :octocat: 👍

File: app/controllers/lessons_controller.rb (hunk 2 above)
Comment (McEileen) [inline app/controllers/lessons_controller.rb]: Hi, I'm looking through your review during learn day (monthly day where we get free time to work on our own projects at my office) and I agree with the refactors you suggested.  I opened issue #73 and assigned myself to it, but I will work on this PR more before working on issue #73.  Thanks again!

**Comment by TheJhyde** [comment]
Looks good to me. I've merged the changes.
